internal/docstore/docstore.go

```diff
@@ -179,21 +195,44 @@ func (l *ActionList) Update(doc Document, mods Mods) *ActionList {
 // See ActionList.Update.
 type Mods map[FieldPath]interface{}
 
-// Do executes the action list. If all the actions executed successfully, Do returns
-// (number of actions, nil). If any failed, Do returns the number of successful
-// actions and an error. In general there is no way to know which actions succeeded,
-// but the error will contain as much information as possible about the failures.
-func (l *ActionList) Do(ctx context.Context) (int, error) {
+// An ActionListError is returned by ActionList.Do. It contains all the errors
+// encountered while executing the ActionList, and the positions of the corresponding
+// actions.
+type ActionListError = driver.ActionListError
+
+// Do executes the action list.
+// In ordered mode (when the Unordered method was not called on the list), execution
+// will stop after the first action that fails.
+// In unordered mode, all the actions will be executed. In either mode, as a special
+// case, none of the actions will be executed if any is invalid (for example, a Put
+// whose document is missing its key field).
+// If any action fails, the returned error will be an ActionListError that contains
```

Comment:
Make it clear if this comment is only for Unordered mode, or for both (again, if it applies to both, consider moving it earlier?).

Reply:
Done.